#301 Add cloning, tests, route

andymikulski opened the pull request:
Fixes #301 (again)

- Add 'clone' button to existing recipes
- Add 'clone' route (e.g. `../recipe/14/clone/`)
- Add clone tests

andymikulski commented:
@Osmose @mythmon r?

mythmon commented on client/control/components/RecipeForm.js:
```diff
@@ -57,8 +99,14 @@ export class RecipeForm extends React.Component {
+
     return (
       <form className="recipe-form" onSubmit={handleSubmit}>
+        { this.renderCloningMessage() }
```
This seems like an odd place to put this. It gives a really high visual priority, since it is the first thing a user will see. On the other hand, it makes sense to have it at the top of the page. Maybe the top right of the page would be better?

We could also punt on the design issue, if you'd rather do a more comprehensive redesign along with the peer approval features.

andymikulski replied:
Ideally, the button would go alongside the 'preview' and 'history' buttons, though I was having trouble displaying that button dynamically up there. I wasn't too concerned with the design at this point, but I can take care of it.

andymikulski commented on client/control/tests/components/test_RecipeForm.js:
```diff
@@ -112,11 +141,32 @@ describe('<RecipeForm>', () => {
     });
 
-    it("should add a new recipe if it doesn't have an ID", async () => {
+    it('should add a new recipe if it doesn\'t have an ID', async () => {
```
Weird, yeah I'm wondering if that was a result of running `lint --fix`

andymikulski commented on client/control/components/RecipeForm.js:
```diff
@@ -37,14 +42,51 @@ export class RecipeForm extends React.Component {
+
+    return isCloning ?
+      (<span className="cloning-message">
+        You're cloning {displayedRecipe.name} ({displayedRecipe.action}).
```
Not sure - I'm wondering if it's how I'm splitting up the lines. Easiest fix is to just switch to `You are`, that work?

andymikulski commented:
aight, since last review:

- Moved 'clone' / 'cancel' buttons to the top navbar (next to 'preview', 'history', etc)
- Corrected bug in Header where relative path links would lead to 404
- Added test for the absolutePath util that Header now uses
- Added 'callout' styles for the `You are now cloning...` message, moved it to the right side of the page*

* This 'design' is most definitely temporary, so I didn't spend much time tweaking it

@mythmon this is ready for another review (minus the `route: pt.object` propType deal)

mythmon commented on client/control/components/RecipeForm.js:
```diff
@@ -37,14 +42,51 @@ export class RecipeForm extends React.Component {
     }),
-  }
+    // route prop passed from router
+    route: pt.object,
```
I didn't realize that `route` was coming from the router directly. I assumed that because this wasn't a connected component that it wasn't at the top level, but I guess I was wrong.

mythmon commented on client/control/components/Header.js:
```diff
@@ -2,11 +2,13 @@ import React, { PropTypes as pt } from 'react';
 import { Link } from 'react-router';
 import Breadcrumbs from 'react-breadcrumbs';
 
+import absolutePath from '../../utils/absolute-path';
```
I find some humor in using a relative import path to import something called "absolute-path". (this is obviously the right think, just amusing)
